app/delegate-server/Contract/PlaceBid.purs

```diff
@@ -0,0 +1,247 @@
+    auctionTerms@(AuctionTerms auctionTermsRec) = auctionInfoRec.auctionTerms
+
+  networkId <- lift getNetworkId
+
+  -- Check that the current time is within the bidding period:
+  nowTime <- lift currentTime
+  when (nowTime < auctionTermsRec.biddingStart) $
+    throwError PlaceBidL2_Error_CurrentTimeBeforeBiddingStart
+  when (nowTime >= auctionTermsRec.biddingEnd) $
+    throwError PlaceBidL2_Error_CurrentTimeAfterBiddingEnd
+
+  -- Get collateral utxo:
+  collateralUtxo <- findCollateralUtxo utxos
+    !? PlaceBidL2_Error_CouldNotFindCollateralUtxo
+
+  -- Get current standing bid utxo:
+  standingBidUtxo /\ oldBidState <- findStandingBidUtxo auctionInfoRec utxos
+    ?? PlaceBidL2_Error_CouldNotFindCurrentStandingBidUtxo
+
+  -- Check bid state transition:
+  let newBidState = wrap $ Just bidTerms
+  success <- liftEffect $ validateNewBid auctionCs auctionTerms oldBidState newBidState
+  unless success $ throwError PlaceBidL2_Error_InvalidBidStateTransition
+
+  -- Build validators:
+  -- TODO: pre-build standing bid validator
+  validators <-
+    withExceptT PlaceBidL2_Error_CouldNotBuildAuctionValidators $
+      mkAuctionValidators auctionCs auctionTerms
+
+  let
+    validatorHashes = unwrap $ validatorHash <$> validators
+
+    txValidRange :: POSIXTimeRange
+    txValidRange = to $ auctionTermsRec.biddingEnd - wrap (BigInt.fromInt 1000)
+
+    spendableUtxos :: UtxoMap
+    spendableUtxos = Map.fromFoldable [ collateralUtxo, standingBidUtxo ]
+
+    -- StandingBid ---------------------------------------------------
+
+    standingBidOref :: TransactionInput
+    standingBidOref = fst standingBidUtxo
+
+    standingBidRedeemer :: Redeemer
+    standingBidRedeemer = wrap $ toData NewBidRedeemer
+
+    standingBidDatum :: Datum
+    standingBidDatum = wrap $ toData newBidState
+
+    standingBidTokenValue :: Value
+    standingBidTokenValue = Value.singleton auctionCs standingBidTokenName one
+
+    --
+
+    balancerConstraints :: BalanceTxConstraintsBuilder
+    balancerConstraints = mconcat
+      [ BalancerConstraints.mustUseCoinSelectionStrategy SelectionStrategyMinimal
+      , BalancerConstraints.mustUseUtxosAtAddresses networkId []
+      , BalancerConstraints.mustUseCollateralUtxos $ Map.fromFoldable [ collateralUtxo ]
+      , BalancerConstraints.mustUseAdditionalUtxos spendableUtxos
+      ]
+
+    constraints :: TxConstraints
+    constraints = mconcat
+      [ -- Spend standing bid utxo with old standing bid datum: 
+        Constraints.mustSpendScriptOutput standingBidOref standingBidRedeemer
+
+      , -- Lock standing bid token with new standing bid datum at
+        -- standing bid validator address:
+        Constraints.mustPayToScript validatorHashes.standingBid standingBidDatum DatumInline
+          standingBidTokenValue
+
+      , -- Set transaction validity interval to bidding period: 
+        Constraints.mustValidateIn txValidRange
+
+      , Constraints.mustNotBeValid
```

Comment:
What's the reason we expect scripts to fail during the balancing?

Reply:
Blockfrost 500 internal error :/

Reply:
Good to know such a constraint exists.